File: VB/T333639/SampleReport.vb
Imports DevExpress.XtraReports.UI
Imports DevExpress.XtraReports.Parameters

Namespace T333639
    Partial Public Class SampleReport
        Inherits DevExpress.XtraReports.UI.XtraReport

        Private _reportColumns As LookUpValueCollection

        Public Sub New()
            InitializeComponent()

            ' Initialize a list of lookup values for report columns.
            _reportColumns = New LookUpValueCollection() From {
                New LookUpValue("ID", "ID"),
                New LookUpValue("Name", "Name"),
                New LookUpValue("Category", "Category"),
                New LookUpValue("Supplier", "Supplier"),
                New LookUpValue("UnitPrice", "Price"),
                New LookUpValue("UnitsInStock", "In Stock"),
                New LookUpValue("UnitsOnOrder", "On Order"),
                New LookUpValue("QuantityPerUnit", "Quantity Per Unit")
            }

            Dim reportColumnsLookUpSettings As New StaticListLookUpSettings()
            reportColumnsLookUpSettings.LookUpValues.AddRange(_reportColumns)

            ' Assign the list of lookup values to the report's ReportColumns parameter.
            Me.Parameters("ReportColumns").LookUpSettings = reportColumnsLookUpSettings
            ' Specify default lookup values.
            Me.Parameters("ReportColumns").Value = New String() { "ID", "Name", "Category", "Supplier", "UnitPrice" }
        End Sub

        Private Sub SampleReport_DataSourceDemanded(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.DataSourceDemanded
            ' Get the selected report columns.
            Dim columns() As String = CType(Me.Parameters("ReportColumns").Value, String())

            Me.tblHeader.BeginInit()
            Me.tblDetail.BeginInit()
            Me.trHeader.Cells.Clear()
            Me.trDetail.Cells.Clear()

            ' Create a table header cell and table body cell for each selected column.
            For Each column As String In columns
                Dim description As String = _reportColumns.Single(Function(lookupValue) lookupValue.Value.Equals(column)).Description

                ' Assign a column description to the table header cell.
                Dim headerCell As New XRTableCell() With {.Text = description}
                Me.trHeader.Cells.Add(headerCell)

                ' Bind the table body cell to the column.
                Dim detailCell As New XRTableCell()
                detailCell.DataBindings.Add("Text", Nothing, String.Format("{0}.{1}", Me.DataMember, column))
                Me.trDetail.Cells.Add(detailCell)
            Next column

            Me.tblHeader.EndInit()
            Me.tblDetail.EndInit()
        End Sub

    End Class
End Namespace

File: CS/T333639/SampleReport.cs
using System;
using System.Drawing;
using System.Collections;
using System.ComponentModel;
using DevExpress.XtraReports.UI;
using DevExpress.XtraReports.Parameters;
using System.Linq;

namespace T333639 {
    public partial class SampleReport : DevExpress.XtraReports.UI.XtraReport {
        LookUpValueCollection _reportColumns;

        public SampleReport() {
            InitializeComponent();

            _reportColumns = new LookUpValueCollection() {
                    new LookUpValue("ID", "ID"),
                    new LookUpValue("Name", "Name"),
                    new LookUpValue("Category", "Category"),
                    new LookUpValue("Supplier", "Supplier"),
                    new LookUpValue("UnitPrice", "Price"),
                    new LookUpValue("UnitsInStock", "In Stock"),
                    new LookUpValue("UnitsOnOrder", "On Order"),
                    new LookUpValue("QuantityPerUnit", "Quantity Per Unit")
            };
            StaticListLookUpSettings reportColumnsLookUpSettings = new StaticListLookUpSettings();
            reportColumnsLookUpSettings.LookUpValues.AddRange(_reportColumns);
            this.Parameters["ReportColumns"].LookUpSettings = reportColumnsLookUpSettings;
            this.Parameters["ReportColumns"].Value = new string[] { "ID", "Name", "Category", "Supplier", "UnitPrice" };
        }

        private void SampleReport_DataSourceDemanded(object sender, EventArgs e) {
            string[] columns = (string[])this.Parameters["ReportColumns"].Value;
            
            this.tblHeader.BeginInit();
            this.tblDetail.BeginInit();
            this.trHeader.Cells.Clear();
            this.trDetail.Cells.Clear();
            foreach (string column in columns) {
                string description = _reportColumns.Single(lookupValue => lookupValue.Value.Equals(column)).Description;

                XRTableCell headerCell = new XRTableCell() {
                    Text = description
                };
                this.trHeader.Cells.Add(headerCell);

                XRTableCell detailCell = new XRTableCell();
                detailCell.DataBindings.Add("Text", null, String.Format("{0}.{1}", this.DataMember, column));
                this.trDetail.Cells.Add(detailCell);
            }
            this.tblHeader.EndInit();
            this.tblDetail.EndInit();
        }

    }
}
